src/mono/wasm/runtime/startup.ts

```diff
@@ -457,20 +465,37 @@ async function instantiate_wasm_module(
 ): Promise<void> {
     // this is called so early that even Module exports like addRunDependency don't exist yet
     try {
-        replace_linker_placeholders(imports, export_linker());
+        let memoryBytes: ArrayBuffer | undefined = undefined;
         await mono_wasm_load_config(Module.configSrc);
         if (runtimeHelpers.diagnosticTracing) console.debug("MONO_WASM: instantiate_wasm_module");
         const assetToLoad = resolve_asset_path("dotnetwasm");
+
+        if (runtimeHelpers.config.cacheMemory && config.hash) {
+            memoryBytes = await getMemory(runtimeHelpers.configurationHash);
+            runtimeHelpers.memoryIsLoaded = !!memoryBytes;
+        }
+        beforeInstantiateWasm.promise_control.resolve();
+
         // FIXME: this would not apply re-try (on connection reset during download) for dotnet.wasm because we could not download the buffer before we pass it to instantiate_wasm_asset
         await start_asset_download(assetToLoad);
         await beforePreInit.promise;
         Module.addRunDependency("instantiate_wasm_module");
+
+        replace_linker_placeholders(imports, export_linker());
         await instantiate_wasm_asset(assetToLoad, imports, successCallback);
         assetToLoad.pendingDownloadInternal = null as any; // GC
         assetToLoad.pendingDownload = null as any; // GC
         assetToLoad.buffer = null as any; // GC
-
         if (runtimeHelpers.diagnosticTracing) console.debug("MONO_WASM: instantiate_wasm_module done");
+
+        if (runtimeHelpers.memoryIsLoaded) {
+            const wasmMemory = anyModule.asm.memory;
+            // .grow() takes a delta compared to the previous size
+            wasmMemory.grow((memoryBytes!.byteLength - wasmMemory.buffer.byteLength + 65535) >>> 16);
+            runtimeHelpers.updateGlobalBufferAndViews(wasmMemory.buffer);
```

Comment:
@radekdoulik I will need to tell emscripten that I resized the memory. 
So the `replacements.updateGlobalBufferAndViews` should survive you PR in some form.